js/src/admin/utils/ExtensionData.js

```diff
@@ -0,0 +1,34 @@
+import ItemList from '../../common/utils/ItemList';
+import listItems from '../../common/helpers/listItems';
+
+export default class ExtensionData {
+  constructor(extension) {
+    this.extension = extension;
+    app.extensionData[extension] = app.extensionData[extension] || {};
+  }
+
+  registerData(type, content, permissionType = null, priority = 0) {
+    if (content instanceof ItemList) {
+      content = listItems(content.toArray());
+    }
+
+    app.extensionData[this.extension][type] = app.extensionData[this.extension][type] || {};
+
+    if (type === 'permissions') {
+      if (!app.extensionData[this.extension][type][permissionType]) {
+        app.extensionData[this.extension][type][permissionType] = new ItemList();
+      }
+      app.extensionData[this.extension][type][permissionType].add(content.permission, content);
+    } else {
+      app.extensionData[this.extension][type] = content;
+    }
+
+    return this;
+  }
+
+  registerPage(component) {
```

Comment:
Hmm, should we store this as a registered route, or perhaps we could just maintain a map of extension IDs to custom page components. That could work just as well for the route resolver, and would minimize any overhead in routes